rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+2. The second drawback is related to the fact that structs take up more storage slots than they actually need. This behavior is **not** new. There is a tradeoff here to be considered. If we make structs tightly packed, then manipulating a single field becomes more expensive (potential read-modify-write patterns or fields "spilling" over to the next slot), but the number of storage slots required may become lower. We might want to consider a `#[packed]` attribute in the future to let the user decide on one way v.s. the other. Another thing to consider with tightly-packed structs is that the type `StorageHandler` would then also require an additional field that points to a particular location _inside_ a given slot (i.e. which byte to start reading from or writing to).
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+One alternative is to continue with the current use model which has many flaws so that's not ideal. Another alternative is to introduce an `AsStorageHandler` trait that allows calling `storage.x.as_storage_handler()` to get the storage handler manually, but that approach has its limitations when implementing `StorageMap` and `StorageVec` and requires defining the behavior of `as_storage_hander` when called on a regular stack variable.
+
+One more alternative is to completely remove the concept of a "storage variable" and switch over to thinking about storage like file I/O or databases. However, I think the approach described in this RFC provides a good middle ground where methods like `store` and `get` somewhat behave like I/O methods but also keeps the abstraction of "storage variables" to help user write readable code.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+Solidity's approach to storage is completely different from the above and has its own problems, as storage variable are hard to identify and storage accesses are not explicit enough. Sway's current approach is also flawed as explained in the [Motivation](#motivation) section.
+
+In Rust, the closest thing to storage accesses is file I/O where a handler is used to read and write to a file similarly to `store` and `get` above.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How do we make storage accesses, particularly reads, more ergonomic?
+- Is not having structs tightly packed acceptable? Is that something that we want in the future, potentially via a `#[packed]` attribute that also reorders the fields of the struct for optimal layout?
+- How do we implement nested dynamic storage types such as `StorageVec<StorageVec<u64>>`?
+- How do we obtain a `StorageHandle` for data stored in a dynamic storage collection such as `StorageMap` or `StorageVec`? Is that even needed/required?
```

Comment:
This could potentially be quite important, especially for collections of large types?

E.g. Given a `StorageVec<VeryLargeType>`, it would be much cheaper for `StorageVec::get` to return a `StorageHandle<VeryLargeType>` than it would be to return the `VeryLargeType` itself. E.g. given a storage handle, the user could potentially defer the cost of reading the `VeryLargeType` based on some condition in the code.

Perhaps `StorageMap::get` and `StorageVec::get` are better off returning `StorageHandler<T>` rather than `Option<T>`? If we're to go for the suggested `read`/`write` methods for differentiation, usage might look like:
```rust
let elem_handle = storage.my_vec.get(ix); // Get the handle.
```
```rust
let elem_opt = storage.my_vec.get(ix).read(); // Read the `Option<T>`.
```

Reply:
I was actually wondering if we need a separate API to get the handler instead of changing the behavior of `get` since `get` is modelled after the `get` method from Rust's `Vec` and `HashMap`.

Maybe something like this would work:
```rust
let elem_handle = storage.my_vec.get_handle(ix); // or `get_key(ix)`, Get the handle
```
```rust
let elem_opt = storage.my_vec.get(ix); // Read the `Option<T>`
```

Reply:
Went with your suggestion here as well.